run_matrix.py

```diff
@@ -0,0 +1,161 @@
+    charm_config: _CharmTestConfig,
+    interface: str,
+    root: Path = Path("/tmp/charm-relation-interfaces-tests/"),
+):
+    """Clone the charm repository and create the venv if it hasn't been done already."""
+    charm_path = root / Path(charm_config.name)
+    if not charm_path.is_dir():
+        branch_option = ""
+        if charm_config.branch:
+            branch_option = f"--branch {charm_config.branch}"
+            logging.warning(f"custom branch provided for {charm_config.name}; this should only be done in staging")
+        subprocess.call(
+            f"git clone --quiet --depth 1 {branch_option} {charm_config.url} {charm_path}",
+            shell=True,
+            stdout=subprocess.DEVNULL
+        )
+        _setup_venv(charm_path)
+    fixture_path, fixture_id = _get_fixture(charm_config, charm_path)
+    if not fixture_path.is_file():
+        # NOTE: In the future we could probably run the tests without a fixture, assuming
+        # that the charm needs no patching at all to work with scenario
+        raise SetupError(f"fixture missing for charm {charm_config.name}")
+    test_path = _generate_test(interface, fixture_path.parent, fixture_id)
+    return charm_path, test_path
+
+
+def _clean(root: Path = Path("/tmp/charm-relation-interfaces-tests/")):
+    """Clean the directory used to store repos for the tests."""
+    if root.is_dir():
+        shutil.rmtree(root)
+
+
+def _generate_test(interface: str, test_path: Path, fixture_id: str) -> Path:
+    """Generate a pytest file for a given charm and interface."""
+    test_content = textwrap.dedent(f"""from interface_tester import InterfaceTester
+    def test_{interface}_interface({fixture_id}: InterfaceTester):
+        {fixture_id}.configure(
+        interface_name="{interface}",
+    )
+    {fixture_id}.run()
+    """)
+    test_filename = f"interface-test-{interface}.py"
+    with open(test_path / test_filename, "w") as file:
+        file.write(test_content)
+    return test_path / test_filename
+
+
+def _get_fixture(charm_config: _CharmTestConfig, charm_path: Path):
+    """Get the tester fixture from a charm."""
+    fixture_path = charm_path / FIXTURE_PATH
+    fixture_id = FIXTURE_IDENTIFIER
+    if charm_config.test_setup:
+        if charm_config.test_setup["location"]:
+            fixture_path = charm_path / Path(charm_config.test_setup["location"])
+        if charm_config.test_setup["identifier"]:
+            fixture_id = charm_config.test_setup["identifier"]
+    return fixture_path, fixture_id
+
+
+def _setup_venv(charm_path: Path) -> Path:
+    """Create the venv for a charm and return the path to its python."""
+    logging.info(f"Installing dependencies in venv for {charm_path}")
+    original_wd = os.getcwd()
+    os.chdir(charm_path)
+    # Create the venv and install the requirements
+    try:
+        subprocess.check_call("python -m venv ./.interface-venv", shell=True, stdout=subprocess.DEVNULL)
+        subprocess.check_call(
```

Comment:
add a note/TODO here that we're not addressing yet charm-specific dependencies. We should eventually do something like `pip install .` (if toml build system is used) or `pip install -r requirements.txt`, or Scenario will never work if the charm requires fancy extra dependencies.

Reply:
We are doing `pip install -r requirements` actually, just below that line!